# Initialise fluid section defaults; require *COUPLING SURFACE; stop on ARPACK failure in arpackcs
#
#   - fluidsections: initialise material and nconstants; reject unknown subtypes
#   - couplings: require the SURFACE parameter on *COUPLING
#   - arpackcs: stop when dnaupd/dsaupd fail, as arpack does
#
# Found by fuzzing ccx with malformed variants of 518 input decks under
# AddressSanitizer, and by review of the code around each crash.

diff --git a/ccx_2.23/src/arpackcs.c b/ccx_2.23/src/arpackcs.c
index 074f4f1..270afc7 100644
--- a/ccx_2.23/src/arpackcs.c
+++ b/ccx_2.23/src/arpackcs.c
@@ -900,6 +900,7 @@ void arpackcs(double *co, ITG *nk, ITG **konp, ITG **ipkonp, char **lakonp,
       if(info!=0){
 	printf(" *ERROR in d[n,s]aupd: info=%" ITGFORMAT "\n",info);
 	printf("       # of converged eigenvalues=%" ITGFORMAT "\n\n",iparam[4]);
+	if(info<0) FORTRAN(stop,());
       }         
 	
       NNEW(temp_array2,double,nev);
diff --git a/ccx_2.23/src/couplings.f b/ccx_2.23/src/couplings.f
index 7737fab..f9d9217 100644
--- a/ccx_2.23/src/couplings.f
+++ b/ccx_2.23/src/couplings.f
@@ -109,6 +109,7 @@
 !     
       name(1:1)=' '
       irefnode=0
+      jsurf=0
       do i=2,n
         if(textpart(i)(1:8).eq.'REFNODE=') then
           read(textpart(i)(9:18),'(i10)',iostat=istat) irefnode
@@ -186,6 +187,15 @@
      &       "*COUPLING%",ier)
         return
       endif
+!     
+      if(jsurf.eq.0) then
+        write(*,*)
+     &       '*ERROR reading *COUPLING: no SURFACE given'
+        write(*,*) '  '
+        call inputerror(inpc,ipoinpc,iline,
+     &       "*COUPLING%",ier)
+        return
+      endif
 !     
       if(orientation.eq.'                    ') then
         iorientation=0
diff --git a/ccx_2.23/src/fluidsections.f b/ccx_2.23/src/fluidsections.f
index cc63f51..5946d29 100644
--- a/ccx_2.23/src/fluidsections.f
+++ b/ccx_2.23/src/fluidsections.f
@@ -58,6 +58,8 @@
       typename='
      &'
       ipos=0
+      material=' '
+      nconstants=-1
 !     
       do i=2,n
         if(textpart(i)(1:9).eq.'MATERIAL=') then
@@ -486,6 +488,11 @@ C     ndprop=20
           elname='RCVL   '
         else if(typename(15:23).eq.'NONLINEAR')then
           elname='RCVN   '
+        else
+          write(*,*) '*ERROR reading *FLUID SECTION: ',typename
+          write(*,*) '       is an unknown fluid section type'
+          ier=1
+          return
         endif
         ndprop=6
 !     
@@ -504,12 +511,24 @@ C     ndprop=20
         elseif(typename(7:12).eq.'FORCED') then
           elname='VOFO   '
           ndprop=10
+        else
+          write(*,*) '*ERROR reading *FLUID SECTION: ',typename
+          write(*,*) '       is an unknown fluid section type'
+          ier=1
+          return
         endif
 !     
       elseif(typename(1:1).eq.' ') then
         elname='       '
         ndprop=0
       elseif(typename(1:1).eq.'U') then
+        if(nconstants.lt.0) then
+          write(*,*) '*ERROR reading *FLUID SECTION: user element ',
+     &         typename(1:7)
+          write(*,*) '       needs a CONSTANTS parameter'
+          ier=1
+          return
+        endif
         elname(1:7)=typename(1:7)
         ndprop=nconstants
       else
